sway-lib-std/src/storage.sw

```diff
@@ -106,3 +109,150 @@ impl<K, V> StorageMap<K, V> {
         get::<V>(key)
     }
 }
+
+/// A persistant vector struct
+pub struct StorageVec<V> {}
+
+impl<V> StorageVec<V> {
+    /// Appends the value to the end of the vector
+    #[storage(read, write)]
+    pub fn push(self, value: V) {
+        // The length of the vec is stored in the __get_storage_key() slot
+        let len = get::<u64>(__get_storage_key());
+        
+        // Storing the value at the current length index (if this is the first item, starts off at 0)
+        let key = sha256((len, __get_storage_key()));
+        store::<V>(key, value);
+
+        // Incrementing the length
+        store(__get_storage_key(), len + 1);
+    }
+
+    /// Removes the last element of the vector
+    #[storage(read, write)]
+    pub fn pop(self) -> Option<V> {
+        let len = get::<u64>(__get_storage_key());
+        // if the length is 0, there is no item to pop from the vec
+        if len == 0 {
+            return Option::None;
+        }
+    
+        // reduces len by 1, effectively removing the last item in the vec
+        store(__get_storage_key(), len - 1);
+
+        let key = sha256((len - 1, __get_storage_key()));
+        Option::Some::<V>(get::<V>(key))
+    }
+
+    /// Gets the value in the given index
+    #[storage(read)]
+    pub fn get(self, index: u64) -> Option<V> {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to return
+        if len <= index {
+            return Option::None;
+        }
+
+        let key = sha256((index, __get_storage_key()));
+        Option::Some::<V>(get::<V>(key))
+    }
+
+    /// Removes the value in the given index and moves all the values in the following indexes
+    /// Down one index
+    /// WARNING: Expensive for larger vecs
+    #[storage(read, write)]
+    pub fn remove(self, index: u64) -> V {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to remove
+        assert(index <= len);
```

Comment:
If there are 3 items in the vector then the len is 3

E.g.

```
a, b, c
```

If the index to remove is `3` then this assertion would allow the theoretical "d" to be removed (which does not exist) because the indexes are a=0, b=1, c=2.
Therefore the assertion should be index < len, [reference](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1410)

This applies to [swap_remove](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1310) too.